OSDOCS-6534: adding CSI snapshotting (#62557)
openshift/openshift-docs: modules/microshift-storage-device-classes.adoc

// Module included in the following assemblies:
//
// * microshift_storage/volume-snapshots-microshift.adoc

:_content-type: CONCEPT
[id="microshift-storage-device-classes_{context}"]

> == Device classes

Reviewer: ```suggestion
= Device classes
```
See the comment above about using an H1 and using the level offset in the assembly.

> == Device classes

Reviewer 2: I thought we were not going to include discussion of multiple device classes in the thin provisioning instructions. Maybe I'm misremembering? This content isn't wrong, but maybe we can move it to a chapter for managing multiple device classes instead of having it in the snapshot chapter.

Author: No, there was no mention of not including this. Removing some devices was a goal. We did remove raid.

Reviewer 3: Could be my ignorance of adoc, but isn't this file already it's own section for device-classes?

IIRC, the consensus was to provide a simple example lvmd.yaml for snapshotting, but also provide an expanded config like the one here.

Author: Yes, this is an individual module. It can be called in any assembly. 

We have another open docs ticket for "day 2" storage config operations. This could go there once that is established--or in an existing section if it makes more sense. OCP weaves this content into each provider, which isn't relevant here. I don't where else it goes in the workflow for a MicroShift admin?

Reviewer 2: https://docs.google.com/document/d/1Slfy4-oHf7RqvP2tVnr6z6jHXwMHJssdOpw-G4zwUts/edit?disco=AAAA1e6X9sw is where we talked about this on the earlier draft.

Including this module in the thin provisioning instructions is mixing unrelated content. We should move it to its own chapter of the storage book.

Author: It's in the LVMS plugin section in OCP, so I parked it there for the moment. I think we can tie that module in with the other work around advanced config to come in a later PR. Sound good?

> You can configure device classes by adding a `device-classes` array the `lvms.yaml` configuration file. A single device class must be set as the default.

Reviewer 4: @ShaunaDiaz This sentence looks a little jumbled.
Also, is the file supposed to be **lvms.yaml** or **lvmd.yaml** ?

> You can create custom device classes by adding a `device-classes` array to your LVMS configuration. Add the array to the `lvmd.yaml` configuration file. A single device class must be set as the default. You must restart {product-title} for configuration changes to take effect.

Reviewer 3: ```suggestion
You can create custom device classes by adding a `device-classes` array to your LVMS configuration. Add the array to the `/etc/microshift/lvmd.yaml` configuration file. A single device class must be set as the default. You must restart {product-title} for configuration changes to take effect.
```

> You can create custom device classes by adding a `device-classes` array to your LVMS configuration. Add the array to the `/etc/microshift/lvmd.yaml` configuration file. A single device class must be set as the default. You must restart {product-title} for configuration changes to take effect.

Reviewer: Reminder to spell out abbreviations on the first usage in a module

> [NOTE]

Reviewer 5: Probably worthy of an IMPORTANT or WARNING

====
Removing a device class while there are still persistent volumes or `VolumeSnapshotContent` objects connected to that device class breaks both thick and thin provisioning.
====

> Multiple device classes can be defined in the `device-classes` array. These classes can be a mix of thick and thin volume configurations.

Reviewer 5: ```suggestion
You can define multiple device classes in the `device-classes` array. These classes can be a mix of thick and thin volume configurations.
```

> .Example of a mixed `device class` array

Reviewer 5: ```suggestion
.Example of a mixed `device-class` array
```

> .Example of a mixed `device class` array
> 
> [source, terminal]

Reviewer: ```suggestion
.Example of a mixed `device class` array
[source, terminal]
```

----
socket-name: /run/topolvm/lvmd.sock
device-classes:
  - name: ssd
    volume-group: ssd-vg
    spare-gb: 0 <1>
    default: true
  - name: hdd
    volume-group: hdd-vg
    spare-gb: 0 <1>
  - name: thin
    spare-gb: 0 <1>
    thin-pool:
      name: thin
      overprovision-ratio: 10
    type: thin
    volume-group: ssd
  - name: striped
    volume-group: multi-pv-vg
    spare-gb: 0 <1>
    stripe: 2
    stripe-size: "64"
    lvcreate-options:<2>
----
[.small]

> <1> Setting spare-gb to anything other than `0` usually results in more space being allocated than expected.

Reviewer: Using the same number for multiple callouts breaks on the Customer Portal. Unfortunately, we need to number sequentially and repeat the same text for each callout.

Reviewer 5: Does it? That's news to me @michaelryanpeter. Do we have it in the guidelines anywhere?

I'd put `spare-gb` in backticks and give it a noun.

Also, is it really only "usually" or will it always allocate more space if you set it?

Could potentially reword to more actionable. First thoughts below, but probably could be better

```suggestion
<1> Set the `spare-gb` field to a value greater than `0` to allocate additional space.
```

Reviewer: @bergerhoffer I think the guidelines could use an update, but here is a rejected Jira asking for support from DXP: https://issues.redhat.com/browse/CCS-6666

Reviewer: An example in the wild: https://access.redhat.com/documentation/en-us/openshift_container_platform/4.12/html/installing/installing-on-aws#installation-aws-config-yaml_installing-aws-customizations

Reviewer 5: Thanks for the example! I'm not surprised how bad that comes out with how ridiculous those conditionals are. I thought I had seen a basic repeater one work on the CP before, but that's hard to find one of them. Sounds like if it's not going to work then we should definitely get it into the guidelines.

> <2> Extra arguments to pass to lvcreate, such as `--type=`. Neither {product-title} nor the LVMS verifies `lvcreate-options`. These option values are passed as-is to the `lvcreate` command. Take special care to ensure that the options specified here are correct.

Reviewer: ```suggestion
<2> Extra arguments to pass to lvcreate, such as `--type=` argument. Neither {product-title} nor the LVMS verifies `lvcreate-options` argument/command/flag/value. These option values are passed as-is to the `lvcreate` command. Take special care to ensure that the options specified here are correct.
```
